File: Maximum Sum Combination[25-September-2023]/Solution.java
//   >> JAVA CODE <<

class Solution {
    static List<Integer> maxCombinations(int N, int K, int A[], int B[]) {
        Arrays.sort(A);
        Arrays.sort(B);
        
        List<Integer> ans=new ArrayList<>();
        
        PriorityQueue<int[]> pq=new PriorityQueue<>((a,b)->{
            if(a[0]==b[0]) return b[1]-a[1];
            
            return b[0]-a[0];
        });
        
        
        for(int i=0;i<N;i++)
        {
          pq.add(new int[]{A[i]+B[N-1],N-1});  
        }
        
        while(!pq.isEmpty() && K-->0)
        {
            int sum=pq.peek()[0];
            int ind=pq.peek()[1];
            pq.poll();
            
            ans.add(sum);
            
            if(ind-1>=0)
                pq.add(new int[]{sum- B[ind]+B[ind-1],ind-1});
        }
        
        return ans;
    }
}

File: Maximum Sum Combination[25-September-2023]/Solution.cpp
//   >> CPP CODE <<

class Solution {
  public:
    vector<int> maxCombinations(int N, int K, vector<int> &A, vector<int> &B) {
        
        sort(A.begin(),A.end());
        sort(B.begin(),B.end());
        
        vector<int>ans;
        priority_queue<pair<int,int>>pq;
        
        for(int i=0;i<N;i++)
        {
            int sum = A[i]+B[N-1];
            pq.push({sum,N-1});
        }
        
        while(!pq.empty() and K--)
        {
            int sum = pq.top().first;
            int ind = pq.top().second;
            pq.pop();
            ans.push_back(sum);
            if(ind-1>=0)
            pq.push({sum-B[ind]+B[ind-1], ind-1});
        }
        
        
        return ans;
    }
};